### PR TITLE
DAOS-10885 control: Suggested updates to GetMaxPoolSize

### PR DESCRIPTION
**Pull request by tanabarr**
Required-githooks: true

Signed-off-by: Tom Nabarro <tom.nabarro@intel.com>

### CODE REVIEW

File: src/control/lib/control/pool.go [hunk 1]
```diff
@@ -1227,105 +1227,139 @@ func ListPools(ctx context.Context, rpcClient UnaryInvoker, req *ListPoolsReq) (
...
+// Add namespace ranks to rankNVMeFreeSpace map and return minimum free available SCM namespace bytes.
+func processSCMSpaceStats(log logging.Logger, checkRank checkRankFn, scmNamespaces storage.ScmNamespaces, rankNVMeFreeSpace rankFreeSpaceMap) (uint64, error) {
+	scmBytes := uint64(math.MaxUint64)
 
-		if hostStorage.ScmNamespaces.Free() == uint64(0) {
-			return 0, 0, errors.Errorf("Host without SCM storage: hostname=%s",
-				hostStorageSet.HostSet.String())
+	for _, scmNamespace := range scmNamespaces {
+		if scmNamespace.Mount == nil {
+			// FIXME: Under what circumstances is this valid, should this
```
Comment (knard-intel) [inline src/control/lib/control/pool.go:1246]: From my side, not enough knowledge of the DAOS internals to define if such situation is managed or if an error should be raised.

File: src/control/lib/control/pool.go [hunk 2]
```diff
@@ -1227,105 +1227,139 @@ func ListPools(ctx context.Context, rpcClient UnaryInvoker, req *ListPoolsReq) (
...
+		sBytes, err := processSCMSpaceStats(log, checkRank, hostStorage.ScmNamespaces, rankNVMeFreeSpace)
+		if err != nil {
+			return 0, 0, err
+		}
+
+		if scmBytes > sBytes {
+			scmBytes = sBytes
+		}
+
+		// FIXME: Should we fail here if any of the namespaces have no free space
```
Comment (knard-intel) [inline src/control/lib/control/pool.go:1344]: This case is managed by the dmg command itself
https://github.com/daos-stack/daos/blob/7440dde89bf57ad2ea040d48bc18c44b09627488/src/control/cmd/dmg/pool.go#L150

As this function is also used by `dmg system storage usage`, an error should not be send at this level.

**Review by knard-intel — COMMENTED** [review]
Your solution is indeed more golang idiomatic and I also prefer the solution of constructing the nvme available size at each step instead of using a marking board: a little bit less efficient but more easy to understand and thus more easy to maintain (which is the most important from my side for such par of the code with no real performance constraint).

I have looked at the test errors and they are not related to your modification (from what I understand): it seems to be more a CI issue.

There is still one pending FIXME which we should define what to do.

A the end if it is OK for you I will cherry-pick your proposed modification and if needed do some slight modification according to CI and reviewers feedbacks. I will also probably add some comments and if needed other minor modifications.

File: src/control/lib/control/pool.go (hunk 2 above)
Comment (tanabarr) [reply src/control/lib/control/pool.go:1344]: ok, let's remove this comment then

File: src/control/lib/control/pool.go (hunk 1 above)
Comment (tanabarr) [reply src/control/lib/control/pool.go:1246]: I think we should make this an error, if not it will give us obscure behaviour that may be hard to diagnose

Comment (knard-intel) [reply src/control/lib/control/pool.go:1246]: OK I will do it.

**Pull request closed without merging**